File: cpp/031-040/Combination Sum II.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <stack>
#include <queue>

using namespace std;

class Solution {
    
    void help(vector<int>& now, vector<int>&candidates, int index, int target, vector<vector<int>>& results)
    {
        for(int i = index;i < candidates.size();i++)
        {
            if(candidates[i] < target)
            {
                now.push_back(candidates[i]);
                help(now, candidates, i+1, target-candidates[i], results);
                now.pop_back();
                while(i < candidates.size()-1 && candidates[i] == candidates[i+1]) i++;
            } else if(candidates[i] == target) {
                now.push_back(candidates[i]);
                results.push_back(now);
                now.pop_back();
                return;
            }
        }
    }
    
public:
    vector<vector<int> > combinationSum2(vector<int> &candidates, int target) {
        vector<vector<int>> results;
        vector<int> tmp;
        sort(candidates.begin(), candidates.end());
        help(tmp, candidates, 0, target, results);
        return results;
    }
    
};

int main()
{
    Solution s;

    return 0;
}

File: python/031-040/Combination Sum II.py
class Solution:
    # @param {integer[]} candidates
    # @param {integer} target
    # @return {integer[][]}
    def combinationSum2(self, candidates, target):
        candidates.sort()
        result = []
        tmp = []
        self.help(tmp, candidates, 0, target, result)
        return result
    
    def help(self, now, candidates, index, target, result):
        i = index
        while i < len(candidates):
            if candidates[i] < target:
                now.append(candidates[i]);
                self.help(now, candidates, i+1, target-candidates[i], result)
                del now[-1]
                while i < len(candidates)-1 and candidates[i] == candidates[i+1]:
                    i += 1
            elif candidates[i] == target:
                temp = now[:]
                temp.append(candidates[i])
                result.append(temp)
                return
            i += 1